Fix 3ac1a2f1e4: Game crash due to invalid vehicle type information.

PeterN opened the pull request:
<!--
Commit message:

- Please use Feature / Add / Change / Fix for player-facing changes. E.g.: "Feature: My cool new feature".
- Please use Feature / Add / Change / Fix followed by "[NewGRF]" or "[Script]" for moddable changes. E.g.: "Feature: [NewGRF] My cool new NewGRF addition".
- Please use Codechange / Codefix for developer-facing changes. E.g.: "Codefix #1234: Validate against nullptr properly".

See https://github.com/OpenTTD/OpenTTD/blob/master/CODINGSTYLE.md#commit-message for more details.
-->

## Motivation / Problem

Game can crash due to invalid data in stored in `Engine`'s union of `RailVehicleInfo`, `RoadVehicleInfo`, `ShipVehicleInfo` and `AircraftVehicleInfo`.

This likely because in 3ac1a2f1e4, `RailVehicleInfo` contains `RailTypes` which is no longer a simple type, which means use inside a union is problematic.

<!--
Describe here shortly
* For bug fixes:
    * What problem does this solve?
    * If there is already an issue, link the issue, otherwise describe the problem here.
* For features or gameplay changes:
    * What was the motivation to develop this feature?
    * Does this address any problem with the gameplay or interface?
    * Which group of players do you think would enjoy this feature?
-->


## Description

Use `std::variant` instead of `union` for vehicle info, ensuring that the value is properly initialised.

<!--
Describe here shortly
* For bug fixes:
    * How is the problem solved?
* For features or gameplay changes:
    * What does this feature do?
    * How does it improve/solve the situation described under 'motivation'.
-->


## Limitations

Possible performance impact as `std::variant` access always checks the requested type is what is stored.
Probably should rename the variable from `u` while we're here...

<!--
Describe here
* Is the problem solved in all scenarios?
* Is this feature complete? Are there things that could be added in the future?
* Are there things that are intentionally left out?
* Do you know of a bug or corner case that does not work?
-->


## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, game_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

rubidium42 commented on src/engine.cpp:
```diff
@@ -78,23 +78,41 @@ Engine::Engine(VehicleType type, uint16_t local_id)
+				/* Set visual effect to the default value */
+				svi.visual_effect = VE_DEFAULT;
+				/* Ships must have a non-zero acceleration. */
+				svi.acceleration = 1;
```
Won't default member initialisation be easier for the `*Info` objects?

PeterN replied:
Other than needing to move the `VisualEffect` enum, yeah.

LordAro commented on src/engine_base.h:
```diff
@@ -64,12 +64,8 @@ struct Engine : EnginePool::PoolItem<&_engine_pool> {
-		AircraftVehicleInfo air;
-	} u{};
+	/* Vehicle-type specific information, not used for Disaster or Effect vehicles. */
+	std::variant<std::monostate, RailVehicleInfo, RoadVehicleInfo, ShipVehicleInfo, AircraftVehicleInfo> u{};
```
could this be made private? (and renamed)

maybe consider its position in the struct too, for memory purposes

PeterN replied:
Yes. Also gave it a more sensible name. Also moved it to the end of the list of members, which does not affect the object size.

PeterN committed:
Fix 3ac1a2f1e4: Game crash due to invalid vehicle type information.

Use std::variant instead of union for vehicle info.

RailVehicleInfo is now non-POD so using in a union causes undefined behaviour.

rubidium42 reviewed (approved):
I entertained the thought of having sub classes of `Engine` like for `Vehicle` and the removal of the internal `EngineInfo` type. Though that only makes our live more difficult when copying data around.
In other words: I can't think of a feasible alternative.
